File: src/handlers/auth/signIn.ts
import { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda';
import middy from '@middy/core';
import httpJsonBodyParser from '@middy/http-json-body-parser';
import httpErrorHandler from '@middy/http-error-handler';
import {
  CognitoIdentityProviderClient,
  InitiateAuthCommand,
  NotAuthorizedException,
  UserNotConfirmedException,
} from '@aws-sdk/client-cognito-identity-provider';
import { z } from 'zod';
import { logger } from '../../shared/powertools';
import { config } from '../../shared/config';

const cognito = new CognitoIdentityProviderClient({});

const SignInSchema = z.object({
  email: z.string().email(),
  password: z.string().min(1),
});

const lambdaHandler = async (event: APIGatewayProxyEvent): Promise<APIGatewayProxyResult> => {
  const parseResult = SignInSchema.safeParse(event.body);
  if (!parseResult.success) {
    return res(400, { message: parseResult.error.issues[0].message });
  }

  const { email, password } = parseResult.data;
  logger.info('Sign in attempt', { email });

  try {
    const { AuthenticationResult } = await cognito.send(new InitiateAuthCommand({
      AuthFlow: 'USER_PASSWORD_AUTH',
      ClientId: config.userPoolClientId,
      AuthParameters: {
        USERNAME: email,
        PASSWORD: password,
      },
    }));

    logger.info('Sign in successful', { email });
    return res(200, {
      idToken: AuthenticationResult!.IdToken,
      accessToken: AuthenticationResult!.AccessToken,
      refreshToken: AuthenticationResult!.RefreshToken,
      expiresIn: AuthenticationResult!.ExpiresIn,
    });
  } catch (err) {
    if (err instanceof NotAuthorizedException) {
      return res(401, { message: 'Incorrect email or password.' });
    }
    if (err instanceof UserNotConfirmedException) {
      return res(403, { message: 'Email not confirmed. Please verify your account first.' });
    }
    logger.error('Sign in failed', { email, error: String(err) });
    throw err;
  }
};

export const handler = middy(lambdaHandler)
  .use(httpJsonBodyParser())
  .use(httpErrorHandler());

function res(statusCode: number, body: object): APIGatewayProxyResult {
  return {
    statusCode,
    headers: { 'Access-Control-Allow-Origin': '*', 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  };
}

File: src/handlers/auth/sign_in.py
import json
import re

import boto3
from botocore.exceptions import ClientError

from src.shared.config import config
from src.shared.powertools import logger, tracer

cognito = boto3.client("cognito-idp")

_EMAIL_RE = re.compile(r"^[^@\s]+@[^@\s]+\.[^@\s]+$")


def _res(status_code: int, body: dict) -> dict:
    return {
        "statusCode": status_code,
        "headers": {"Access-Control-Allow-Origin": "*", "Content-Type": "application/json"},
        "body": json.dumps(body),
    }


@logger.inject_lambda_context
@tracer.capture_lambda_handler
def handler(event, context):
    try:
        body = json.loads(event.get("body") or "{}")
    except (json.JSONDecodeError, TypeError):
        return _res(400, {"message": "Invalid JSON body"})

    email = str(body.get("email", "")).strip()
    password = str(body.get("password", ""))

    if not email or not _EMAIL_RE.match(email):
        return _res(400, {"message": "Invalid email address"})
    if not password:
        return _res(400, {"message": "Password is required"})

    logger.info("Sign in attempt", email=email)

    try:
        response = cognito.initiate_auth(
            AuthFlow="USER_PASSWORD_AUTH",
            ClientId=config["user_pool_client_id"],
            AuthParameters={"USERNAME": email, "PASSWORD": password},
        )
        result = response["AuthenticationResult"]
        logger.info("Sign in successful", email=email)
        return _res(200, {
            "idToken": result["IdToken"],
            "accessToken": result["AccessToken"],
            "refreshToken": result["RefreshToken"],
            "expiresIn": result["ExpiresIn"],
        })
    except ClientError as e:
        error_code = e.response["Error"]["Code"]
        if error_code == "NotAuthorizedException":
            return _res(401, {"message": "Incorrect email or password."})
        if error_code == "UserNotConfirmedException":
            return _res(403, {"message": "Email not confirmed. Please verify your account first."})
        logger.error("Sign in failed", email=email, error=str(e))
        raise
